### ISSUE TITLE
Windows NT4: Installation of Service Pack 6 breaks logon

### ISSUE THREAD

**Issue by herrnst** [joncampbell123/dosbox-x#4463]
### Describe the bug

After installation of Service Pack 6a in a freshly installed Windows NT4 guest, logon isn't possible anymore. After bootup, one can send CTRL-ALT-DEL to enter credentials to log on, but after entering the credentials, a dialog pops up saying "Logging on" ("Anmeldung erfolgt..." in german), and after a while an error is shown saying that the logon service isn't running and the user cannot be logged on at this time.

The same NT CD image plus sp6i386.exe was tested with a non-kvm QEMU VM (configured to resemble the environment DOSBox-X provides to the guest), after installation of the service pack, logon is perfectly possible. Booting the (working in QEMU) harddisk image with DOSBox-X then shows the same failure (i.e "logon service not running").

### Steps to reproduce the behaviour

1. Create a NT4 guest install (boot DOS, format virtual disk, run winnt.exe /b, proceed with the installation)
2. Provide the guest with the sp6i386.exe installation file via an ISO file or any other means
3. Install SP6a by running sp6i386.exe
4. Reboot
5. Logon isn't possible anymore

### Expected behavior

Logon is possible after the service pack installation.

### What operating system(s) this bug have occurred on?

Linux (Arch x86_64, SDL2), macOS (ARM/M1, SDL2 via brew)

### What version(s) of DOSBox-X have this bug?

2023.09.01

### Used configuration

```ini
# relevant items:
[dosbox]
memsize=512
machine=svga_s3trio64v+

[cpu]
cputype=ppro_slow
core=dynamic_rec # tried "normal" aswell
cycles=max

[video]
vmemsize=8 # tried 4 w/o change

[ide, {primary,secondary}]
int13fakeio=true
int13fakev86io=false
enable pio32=true

[serial] -> All ports disabled
[parallel] -> All ports incl. dongle disabled
[printer] -> false

[autoexec]
IMGMOUNT 2 hdd.img -t hdd -fs none -ide 1m -reservecyl 1
# IMGMOUNT D sp6.iso -t iso -ide 2m
BOOT C:
```

### Output log

```text
Nothing printed to the log when trying to log on.
```

### Have you checked that no similar bug report(s) exist?

- [X] I have searched and didn't find any similar bug report.

### Code of Conduct & Contributing Guidelines

- [X] I agree to follow the code of conduct and the contributing guidelines.

**Comment by maxpat78** [comment]
After SP6a, I can't even see the login: a BSOD with "inaccessible boot device" pops up in the very early stage.

**Comment by herrnst** [comment]
Well, I managed to get this installed properly once by copying the SP6a installation package/exe from the ISO to the NT user's desktop, then shutdown the emulated system, remove the IMGMOUNT from the .conf, start the machine again (without a virtual CD drive then ofc) and install from here - doing like this gave a working NT4+SP6a install, where then all kinds of other stuff like IE4+shell update, IE6 and so on could be installed without issues.

This didn't help wrt https://github.com/joncampbell123/dosbox-x/issues/4725 though.

**Comment by maxpat78** [comment]
@herrnst It didn't work for me.

Also, I have _not_ used CPU/Turbo mode, and installed to an IMG (not VHD) file, just to be sure: but, after the reboot, I've had again the BSOD with "inaccessible boot device".
The config problem (#4275) could be the cause: I suspect it is related to the way data are committed to disk by DOSBox-X in a MT environment like NT (tough the use of Start/Shutdown should prevent any problem and ensure all disk activity has been stopped), and/or the DOSBox-X shutdown procedure itself.

**Comment by herrnst** [comment]
Of course it might have been pure luck that it worked one time, though I didn't bother to play around with this further as the "old config used" issue is annoying enough to just not want to play with old NT OSes right now...

**Comment by maxpat78** [comment]
Yes, perhaps DBX isn't mature enough to play NT without issues...

**Comment by BridgeHeadland** [comment]
I have found this at https://groups.google.com/g/comp.os.ms-windows.nt.misc/c/CLEOm3UExNI?pli=1 and read a bit. This does not apply to SP6a, but to SP4, but I don't think it matters that much.
It seems to me that Service Pack 6a is not compatible with the type of SCSI controller DOSBox-X emulates, if it emulates any SCSI controller, it doesn't say anything about SCSI in the DOSBox-X configuration.
I have tried the solutions that are written, but they did not work, maybe I did something wrong.

```ini
STOP 0x0000007B INACCESSIBLE_BOOT_DEVICE After Applying Windows NT 4.0
Service Pack 4
ID: Q214810
----------------------------------------------------------------------------
----
The information in this article applies to:

Microsoft Windows NT Server version 4.0
Microsoft Windows NT Workstation version 4.0

----------------------------------------------------------------------------
----

SYMPTOMS
In some instances, after the application of Service Pack 4, when the
computer is restarted, it will do either of the following:

The system will fail to boot and will display the following error:

Stop 0x0000007B INACCESSIBLE_BOOT_DEVICE
The system will boot, but some devices will not be recognized on the SCSI
bus.

CAUSE
This problem occurs because the Aic78xx.sys driver chosen during the
application of Service Pack 4 is not compatible with the computer's Adaptec
SCSI controller.

RESOLUTION
This problem can be corrected by replacing the existing Aic78xx.sys driver,
located in the %systemroot%\Winnt\System32\Drivers folder, with the
Aic78xx.sys driver from the i386 directory within Service Pack 4.

Windows NT 4.0 Service Pack 4 contains the following Aic78xx drivers:

Aic78xx.sys 26,704 bytes
Aic78xx.001 56,272 bytes

To determine which Aic78xx.sys driver is in use, and to copy the correct
driver, depends upon the file system on the Windows NT boot partition (where
%SystemRoot%\Winnt resides).

If Windows NT resides on a FAT formatted partition, perform the following
steps:

Boot to MS-DOS by using an MS-DOS 6.x boot floppy disk.

Change to the %systemroot%\Winnt\System32\Drivers folder.

Compare the size of AIC78xx.sys with the table above.

Replace the existing AIC78xx.sys file in %systemroot%\Winnt\System32\Drivers
with the other AIC78xx file in the i386 directory within Serivce Pack 4.

If AIC78xx.001 is copied, be sure to rename it AIC78xx.sys.

If Windows NT resides on an NTFS formatted partition, perform the following
steps:

Either boot to a parallel installation of Windows NT or create a parallel
installation of Windows NT. To create a parallel copy of Windows NT, perform
a fresh installation of Windows NT to a different folder.

From the parallel installation, look in the
%systemroot%\Winnt\System32\Drivers of the failing copy of Windows NT.

Compare the size of AIC78xx.sys with the table above.

Replace the existing AIC78xx.sys file in %systemroot%\Winnt\System32\Drivers
with the other AIC78xx file in the i386 directory within Serivce Pack 4.

If AIC78xx.001 is copied, be sure to rename it AIC78xx.sys.

MORE INFORMATION
There are two different versions of the AIC78XX driver for the Adaptec 2940
and 7800 family of SCSI controllers included in Service Pack 4. The
Aic78xx.sys file in Service Pack 4 is the same file that is in Service Pack
3. The Aic78xx.001 file of Service Pack 4 is only placed on the computer if
%systemroot%\Winnt\System32\Drivers\AIC78xx.sys has a revision between
x.x.1382.x and x.x.1420.x.

Additional query words: Stop 0x0000007B AHA-2940 nt 4.00 sp4

Keywords : ntstop
Version : winnt:4.0
Platform : winnt
Issue type : kbprb

Last Reviewed: February 16, 1999
```

**Comment by joncampbell123** [comment]
DOSBox-X doesn't emulate SCSI.

Yet.

There are plans to emulate a specific SCSI controller, and then emulate through it a SCSI CD-ROM drive, so that Windows NT 3.1 can install directly from it. Windows NT 3.1 was made before IDE ATAPI CD-ROM drives were a standard, therefore the only IDE support available for it is for hard drives.

**Comment by maxpat78** [comment]
@BridgeHeadland I don't think aic78xx.sys is the cause.

I took registry snapshosts before and after SP6a setup, and obviously that driver seems disabled (Start=4; if DOSBox-X does not emulate SCSI, there was no reason for NT to load it):
```
[HKEY_LOCAL_MACHINE\SYSTEM\ControlSet001\Services\aic78xx]
"Type"=dword:00000001
"Start"=dword:00000004
"Group"="SCSI miniport"
"ErrorControl"=dword:00000001
"Tag"=dword:0000001e
```

Moreover:
- no particular changes were made to services by SP6a SETUP;
- `/BOOTLOG` option added to BOOT.INI didn't produce any log, probably becuase the STOP occurs in the very early stage;
- I don't know how to debug the NT kernel: a WINDBG debugger could be attached to emulated COM port perhaps? Could I change internal DOSBox-X debugger to handle linear addresses with disassembler?
